File: java/Exercises.java
import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.Optional;
import java.util.function.Predicate;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class Exercises {
    static Map<Integer, Long> change(long amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount cannot be negative");
        }
        var counts = new HashMap<Integer, Long>();
        for (var denomination : List.of(25, 10, 5, 1)) {
            counts.put(denomination, amount / denomination);
            amount %= denomination;
        }
        return counts;
    }

    // Write your first then lower case function here
    public static Optional<String> firstThenLowerCase(List<String> list, Predicate<String> predicate) {
        return list.stream()
                .filter(predicate)
                .findFirst()
                .map(String::toLowerCase);
    }

    // Write your say function here
    static class Say {
        private final StringBuilder phrase;

        public Say(String word) {
            this.phrase = new StringBuilder(word);
        }

        public Say and(String word) {
            if (phrase.length() > 0) {
                phrase.append(" ");
            }
            phrase.append(word);
            return this;
        }

        public String phrase() {
            return phrase.toString();
        }


    }

    static Say say(String word) {
        return new Say(word);
    }

    static Say say() {
        return new Say("");
    }

    // Write your line count function here
    public static long meaningfulLineCount(String filename) throws IOException {
        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            return reader.lines()
                    .filter(line -> !line.trim().isEmpty()) // Not empty or whitespace
                    .filter(line -> !line.trim().startsWith("#")) // Does not start with #
                    .count(); // Count valid lines
        }
    }
}

// Write your Quaternion record class here
record Quaternion(double a, double b, double c, double d) {



    public static final Quaternion ZERO = new Quaternion(0, 0, 0, 0);
    public static final Quaternion I = new Quaternion(0, 1, 0, 0);
    public static final Quaternion J = new Quaternion(0, 0, 1, 0);
    public static final Quaternion K = new Quaternion(0, 0, 0, 1);

    public Quaternion {
        if (Double.isNaN(a) || Double.isNaN(b) || Double.isNaN(c) || Double.isNaN(d)){
            throw new IllegalArgumentException("Coefficients cannot be NaN");
        }
    }

    public Quaternion plus(Quaternion other) {
        return new Quaternion(this.a + other.a, this.b + other.b, this.c + other.c, this.d + other.d);
    }

    public Quaternion times(Quaternion other) {
        double newW = this.a * other.a - this.b * other.b - this.c * other.c - this.d * other.d;
        double newX = this.a* other.b + this.b * other.a + this.c * other.d - this.d * other.c;
        double newY = this.a * other.c - this.b * other.d + this.c * other.a + this.d * other.b;
        double newZ = this.a * other.d + this.b * other.c - this.c * other.b + this.d * other.a;
        return new Quaternion(newW, newX, newY, newZ);
    }

    public Quaternion conjugate() {
        return new Quaternion(a, -b, -c, -d);
    }

    public List<Double> coefficients() {
        return List.copyOf(List.of(a,b,c,d));
    }

    @Override
    public String toString() {
        StringBuilder parts = new StringBuilder();

        if (this.a != 0) parts.append(String.format("%.1f", this.a));
        if (this.b != 0) {
            if (parts.length() > 0) parts.append(this.b > 0 ? "+" : "");
            parts.append(this.b == 1 ? "i" : (this.b == -1 ? "-i" : String.format("%.1f", this.b) + "i"));
        }
        if (this.c != 0) {
            if (parts.length() > 0) parts.append(this.c > 0 ? "+" : "");
            parts.append(this.c == 1 ? "j" : (this.c == -1 ? "-j" : String.format("%.1f", this.c) + "j"));
        }
        if (this.d != 0) {
            if (parts.length() > 0) parts.append(this.d > 0 ? "+" : "");
            parts.append(this.d == 1 ? "k" : (this.d == -1 ? "-k" : String.format("%.1f", this.d) + "k"));
        }

        String result = parts.toString();
        result = result.replace("+-", "-");

        return result.isEmpty() ? "0" : result;
    }
    
}

// Write your BinarySearchTree sealed interface and its implementations here
sealed interface BinarySearchTree permits Empty, Node {
    BinarySearchTree insert(String value);
    boolean contains(String value);
    int size();
    String toString();
}

final class Empty implements BinarySearchTree {
    public static final Empty INSTANCE = new Empty();

    private Empty() {}

    @Override
    public BinarySearchTree insert(String value) {
        return new Node(value, this, this);
    }

    @Override
    public boolean contains(String value) {
        return false;
    }

    @Override
    public int size() {
        return 0;
    }

    @Override
    public String toString() {
        return "";
    }
}


final class Node implements BinarySearchTree {
    private final String value;
    private final BinarySearchTree left;
    private final BinarySearchTree right;

    public Node(String value, BinarySearchTree left, BinarySearchTree right) {
        this.value = value;
        this.left = left;
        this.right = right;
    }


    @Override
    public BinarySearchTree insert(String value) {
        if (value.compareTo(this.value) < 0) {
            // Insert into the left subtree
            return new Node(this.value, left.insert(value), right);
        } else if (value.compareTo(this.value) > 0) {
            // Insert into the right subtree
            return new Node(this.value, left, right.insert(value));
        } else {
            // If the value already exists, return the current node (no duplicates)
            return this;
        }
    }

    @Override
    public boolean contains(String value) {
        if (value.equals(this.value)) {
            return true;
        } else if (value.compareTo(this.value) < 0) {
            // Search in the left subtree
            return left.contains(value);
        } else {
            // Search in the right subtree
            return right.contains(value);
        }
    }

    @Override
    public int size() {
        // Size is the current node (1) + size of left + size of right
        return 1 + left.size() + right.size();
    }

    @Override
    public String toString() {
        // Represent the tree structure, using in-order traversal
        return String.format("(%s)%s(%s)", left, value, right);
    }
}

File: kotlin/Exercises.kt
import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException

fun change(amount: Long): Map<Int, Long> {
    require(amount >= 0) { "Amount cannot be negative" }
    
    val counts = mutableMapOf<Int, Long>()
    var remaining = amount
    for (denomination in listOf(25, 10, 5, 1)) {
        counts[denomination] = remaining / denomination
        remaining %= denomination
    }
    return counts
}

// Write your first then lower case function here
fun firstThenLowerCase(a: List<String>, p: (String) -> Boolean): String? {
    return a.firstOrNull(p)?.lowercase()
}

// Write your say function here
class Say(private val words: String = "") {

    // method to add strings
    fun and(nextWord: String): Say {
        val newPhrase = if (words.isEmpty()) nextWord else "$words $nextWord"
        if (nextWord.isEmpty()) return Say(words + " ")
        return Say(newPhrase)
    }

    val phrase: String
        get() = words
}

fun say(initialWord: String = ""): Say {
    return Say(initialWord)
}

// Write your meaningfulLineCount function here
@Throws(IOException::class)
fun meaningfulLineCount(filename: String): Long {
    BufferedReader(FileReader(filename)).use { reader -> 
        return reader.lineSequence()
            .filter { it.isNotBlank() && !it.trimStart().startsWith("#") }
            .count()
            .toLong()
    }
}


// Write your Quaternion data class here
data class Quaternion private constructor(
    val a: Double,
    val b: Double,
    val c: Double,
    val d: Double
) {

    companion object {
        val ZERO = Quaternion(0.0, 0.0, 0.0, 0.0)
        val I = Quaternion(0.0, 1.0, 0.0, 0.0)
        val J = Quaternion(0.0, 0.0, 1.0, 0.0)
        val K = Quaternion(0.0, 0.0, 0.0, 1.0)

        operator fun invoke(a: Double, b: Double, c: Double, d: Double): Quaternion {
            return Quaternion(a, b, c, d)
        }
    }

    // quat addition
    operator fun plus(other: Quaternion): Quaternion {
        return Quaternion(
            this.a + other.a,
            this.b + other.b,
            this.c + other.c,
            this.d + other.d
        )
    }

    // quat multiplication
    operator fun times(other: Quaternion): Quaternion {
        val newA = this.a * other.a - this.b * other.b - this.c * other.c - this.d * other.d
        val newB = this.a * other.b + this.b * other.a + this.c * other.d - this.d * other.c
        val newC = this.a * other.c - this.b * other.d + this.c * other.a + this.d * other.b
        val newD = this.a * other.d + this.b * other.c - this.c * other.b + this.d * other.a
        return Quaternion(newA, newB, newC, newD)
    }

    fun coefficients(): List<Double> {
        return listOf(a, b, c, d)
    }

    // Conjugate
    fun conjugate(): Quaternion {
        return Quaternion(a, -b, -c, -d)
    } 

    override fun toString(): String {
        val sb = StringBuilder()

        if (a != 0.0) sb.append(a)
        
        if (b == 1.0) sb.append(if (sb.isNotEmpty()) "+" else "").append("i")
        else if (b == -1.0) sb.append("-i")
        else if (b != 0.0) sb.append(if (b > 0 && sb.isNotEmpty()) "+" else "").append("${b}i")

        if (c == 1.0) sb.append(if (sb.isNotEmpty()) "+" else "").append("j")
        else if (c == -1.0) sb.append("-j")
        else if (c != 0.0) sb.append(if (c > 0 && sb.isNotEmpty()) "+" else "").append("${c}j")

        if (d == 1.0) sb.append(if (sb.isNotEmpty()) "+" else "").append("k")
        else if (d == -1.0) sb.append("-k")
        else if (d != 0.0) sb.append(if (d > 0 && sb.isNotEmpty()) "+" else "").append("${d}k")
        
        return if (sb.isEmpty()) "0" else sb.toString()
    }
    
}

// Write your Binary Search Tree interface and implementing classes here
sealed interface BinarySearchTree {
    fun size(): Int
    fun contains(value: String): Boolean
    fun insert(value: String): BinarySearchTree

    // To convert tree into String
    override fun toString(): String

    object Empty : BinarySearchTree {
        override fun size(): Int = 0
        override fun contains(value: String): Boolean = false
        override fun insert(value: String): BinarySearchTree = Node(value, Empty, Empty)
        override fun toString(): String = "()"
    }

    data class Node(
        val value: String,
        val left: BinarySearchTree,
        val right: BinarySearchTree
    ) : BinarySearchTree {
        override fun size(): Int = 1 + left.size() + right.size()
        
        // check if tree contains string
        override fun contains(value: String): Boolean {
            return when {
                this.value == value -> true
                value < this.value -> left.contains(value)
                else -> right.contains(value)
            }
        }

        // insert String into tree
        override fun insert(value: String): BinarySearchTree {
            return when {
                value < this.value -> copy(left = left.insert(value))
                value > this.value -> copy(right = right.insert(value))
                else -> this // val exists leave as is
            }
        }
        
        override fun toString(): String {
            val leftStr = if (left == Empty) "" else left.toString()
            val rightStr = if (right == Empty) "" else right.toString()
            return "($leftStr$value$rightStr)"
        }
    }
}
